### PR TITLE
JDK-8327474 Review use of java.io.tmpdir in jdk tests

### PR DESCRIPTION
**Pull request by bwhuang-us**
This task addresses an essential aspect of our testing infrastructure: the proper handling and cleanup of temporary files and socket files created during test execution. The motivation behind these changes is to prevent the accumulation of unnecessary files in the default temporary directory, which can affect the system's storage and potentially influence subsequent test runs.

Our review identified that several tests create temporary files or socket files without ensuring their removal post-execution. 
- Direct calls to java.io.File.createTempFile and java.nio.file.Files.createTempFile without adequate cleanup.
- Tests using NIO socket channels with StandardProtocolFamily.UNIX, not explicitly removing socket files post-use.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327474](https://bugs.openjdk.org/browse/JDK-8327474): Review use of java.io.tmpdir in jdk tests (**Bug** - P4)

### Reviewers
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**) ⚠️ Review applies to [2517f756](https://git.openjdk.org/jdk/pull/18352/files/2517f7569c8f1fffb711cf88959f4b231d578de9)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18352/head:pull/18352` \
`$ git checkout pull/18352`

Update a local copy of the PR: \
`$ git checkout pull/18352` \
`$ git pull https://git.openjdk.org/jdk.git pull/18352/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18352`

View PR using the GUI difftool: \
`$ git pr show -t 18352`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18352.diff">https://git.openjdk.org/jdk/pull/18352.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18352#issuecomment-2004443450)

### CODE REVIEW

File: test/jdk/java/nio/channels/unixdomain/Bind.java [hunk 1]
```diff
@@ -188,6 +188,7 @@ public static void runTests() throws IOException {
             server = ServerSocketChannel.open(StandardProtocolFamily.UNIX);
             server.bind(null);
             UnixDomainSocketAddress usa = (UnixDomainSocketAddress)server.getLocalAddress();
+            usa.getPath().toFile().deleteOnExit();
```
Comment (AlanBateman) [inline test/jdk/java/nio/channels/unixdomain/Bind.java:191]: The test already deletes the file, I think you just want a try-finally here, same comment on a few other tests.

**Comment by egahlin** [comment]
Speaking for JFR, we should probably just create a normal file, possibly with a filename to indicate subtest and iteration. That said, test changes for JFR looks fine, and fixing the filename is outside the scope of this bug.

File: test/jdk/com/sun/management/HotSpotDiagnosticMXBean/CheckOrigin.java [hunk 2]
```diff
@@ -54,6 +54,7 @@ public static void main(String... args) throws Exception {
             // start a process that has options set in a number of different ways
 
             File flagsFile = File.createTempFile("CheckOriginFlags", null);
+            flagsFile.deleteOnExit();
```
Comment (jaikiran) [inline test/jdk/com/sun/management/HotSpotDiagnosticMXBean/CheckOrigin.java:57]: Hello Bill, jtreg uses a scratch directory when running tests. When a test is launched, the current working directory points to the scratch directory for the test that's currently executing. jtreg manages the lifecycle of scratch directories and even cleans them up (as necessary).
Would it instead be better to just create the temporary file within the jtreg scratch directory (represented by the current working directory)? That way you could just do:

```
File flagsFile = Files.createTempFile(Path.of("."), "CheckOriginFlags", null).toFile();
```
and don't need any explicit deletions?

File: test/jdk/java/nio/channels/unixdomain/Bind.java [hunk 3]
```diff
@@ -185,12 +189,17 @@ public static void runTests() throws IOException {
...
+            UnixDomainSocketAddress usa = null;
+            try {
+                server = ServerSocketChannel.open(StandardProtocolFamily.UNIX);
+                server.bind(null);
+                usa = (UnixDomainSocketAddress) server.getLocalAddress();
+                if (usa.getPath().toString().isEmpty())
+                    throw new RuntimeException("expected non zero address length");
+                System.out.println("Null server address: " + server.getLocalAddress());
+            } finally {
+                Files.deleteIfExists(usa.getPath());
```
Comment (jaikiran) [inline test/jdk/java/nio/channels/unixdomain/Bind.java:201]: `usa` can be `null` here, if it never got assigned due to some exception in the prior lines, which can lead to a `NullPointerException` here.

File: test/jdk/java/util/zip/ZipFile/ZeroDate.java [hunk 4]
```diff
@@ -86,7 +92,7 @@ private static void testDate(byte[] data, int date, LocalDateTime expected) thro
         writeU32(data, locpos + LOCTIM, date);
 
         // ensure that the archive is still readable, and the date is 1979-11-30
-        Path path = Files.createTempFile("out", ".zip");
+        Path path = Utils.createTempFile("out", ".zip");
```
Comment (jaikiran) [inline test/jdk/java/util/zip/ZipFile/ZeroDate.java:95]: So it looks like the test library has this utility method which allows to create temporary files within the jtreg scratch directory. Perhaps we should use it then. Having said that, is there a reason why one test method in this test now uses `Files.createTempFile(...)` and this other test method uses `Utils.createTempFile(...)`?

File: test/jdk/com/sun/management/HotSpotDiagnosticMXBean/CheckOrigin.java (hunk 2 above)
Comment (bwhuang-us) [reply test/jdk/com/sun/management/HotSpotDiagnosticMXBean/CheckOrigin.java:57]: Hi Jaikiran, I think both solutions work for this bug. I personally prefer to place the files in the scratch directory for the ease of debugging. 
In addition, for this specific test, I am considering using File.createTempFile("CheckOriginaFlags", null, Path.of(".").toFile) instead of Files.createTempFile for consistency purposes, as Files.createTempFile may have more restrictive access permissions.

File: test/jdk/java/util/zip/ZipFile/ZeroDate.java (hunk 4 above)
Comment (bwhuang-us) [reply test/jdk/java/util/zip/ZipFile/ZeroDate.java:95]: Yes, you are right. We don't need explicit deletion for these files by using this util method, Utils.createTempFile.

**Comment by seanjmullan** [comment]
The bug should have a `noreg-self` label.

File: test/jdk/java/nio/channels/unixdomain/Bind.java [hunk 5]
```diff
@@ -159,13 +159,17 @@ public static void runTests() throws IOException {
         });
         // address with space should work
         checkNormal(() -> {
-            server = ServerSocketChannel.open(StandardProtocolFamily.UNIX);
-            UnixDomainSocketAddress usa =  UnixDomainSocketAddress.of("with space"); // relative to CWD
+            UnixDomainSocketAddress usa =  UnixDomainSocketAddress.of("with space");
```
Comment (turbanoff) [inline test/jdk/java/nio/channels/unixdomain/Bind.java:162]:
```suggestion
            UnixDomainSocketAddress usa = UnixDomainSocketAddress.of("with space");
```

File: test/jdk/java/nio/channels/unixdomain/NonBlockingAccept.java [hunk 6]
```diff
@@ -48,17 +51,21 @@ static void checkSupported() {
...
             serverSocketChannel.configureBlocking(false);
             serverSocketChannel.bind(null);
+            addr = (UnixDomainSocketAddress) serverSocketChannel.getLocalAddress();
             SocketChannel socketChannel = serverSocketChannel.accept();
             System.out.println("The socketChannel is : expected Null " + socketChannel);
             if (socketChannel != null)
                 throw new RuntimeException("expected null");
             // or exception could be thrown otherwise
+        } finally {
+            Files.deleteIfExists(addr.getPath());
```
Comment (jaikiran) [inline test/jdk/java/nio/channels/unixdomain/NonBlockingAccept.java:68]: Hello Bill, `addr` can be `null` here, so this has a potential of a `NullPointerException`.

File: test/jdk/sun/security/pkcs12/P12SecretKey.java [hunk 7]
```diff
@@ -66,6 +66,7 @@ private void run(String keystoreType, String algName, int keySize) throws Except
         ks.setEntry(ALIAS, ske, kspp);
 
         File ksFile = File.createTempFile("test", ".test");
+        ksFile.deleteOnExit();
```
Comment (jaikiran) [inline test/jdk/sun/security/pkcs12/P12SecretKey.java:69]: Is there a reason why this change is done differently than the rest of the changes in this PR? Can we delete the `File` in a try/finally instead of enrolling a shutdown hook to delete it?

**Comment by jaikiran** [comment]
For the changes in `test/jdk/java/nio/file/Files/CheckPermissions.java` it would be good to have Alan and/or Brian's inputs. The test documentation says that the test relates to SecurityManager checks (and not filesystem permission checks), so the proposed changes in this PR to use a different parent directory for the temporary file(s) should be OK, but I am not familiar with that area to be certain about it.

**Review by jaikiran — APPROVED** [review]
Thank you for the update, Bill. The latest changes look OK to me.
I also heard from Bill that Alan has approved the changes too.

Before integrating, please run tier1, tier2 and tier3 (I haven't checked which tiers these changed tests belong to) to make sure nothing breaks unexpectedly.
